File: benchmarks/Lua/queens.lua
-- This code is derived from the SOM benchmarks, see AUTHORS.md file.
--
-- Copyright (c) 2016 Francois Perrad <francois.perrad@gadz.org>
--
-- Permission is hereby granted, free of charge, to any person obtaining a copy
-- of this software and associated documentation files (the 'Software'), to deal
-- in the Software without restriction, including without limitation the rights
-- to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-- copies of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:
--
-- The above copyright notice and this permission notice shall be included in
-- all copies or substantial portions of the Software.
--
-- THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-- THE SOFTWARE.

local queens = {} do
setmetatable(queens, {__index = require'benchmark'})

function queens:benchmark ()
    local result = true
    for _ = 1, 10 do
        result = result and self:queens()
    end
    return result
end

function queens:verify_result (result)
    return result
end

function queens:queens ()
    self.free_rows  = {true, true, true, true, true, true, true, true}
    self.free_maxs  = {true, true, true, true, true, true, true, true,
                       true, true, true, true, true, true, true, true}
    self.free_mins  = {true, true, true, true, true, true, true, true,
                       true, true, true, true, true, true, true, true}
    self.queen_rows = {-1,   -1,   -1,   -1,   -1,   -1,   -1,   -1}
    return self:place_queen(1)
end

function queens:place_queen (c)
    for r = 1, 8 do
        if self:get_row_column(r, c) then
            self.queen_rows[r] = c
            self:set_row_column(r, c, false)
            if c == 8 then
                return true
            end
            if self:place_queen(c + 1) then
                return true
            end
            self:set_row_column(r, c, true)
        end
    end
    return false
end

function queens:get_row_column (r, c)
    return self.free_rows[r] and self.free_maxs[c + r] and self.free_mins[c - r + 8]
end

function queens:set_row_column (r, c, v)
    self.free_rows[r        ] = v
    self.free_maxs[c + r    ] = v
    self.free_mins[c - r + 8] = v
end

end -- object queens

return queens

File: benchmarks/CSharp/Queens.cs
namespace Benchmarks;

public class Queens : Benchmark
{
  private bool[] freeMaxs = Array.Empty<bool>();
  private bool[] freeRows = Array.Empty<bool>();
  private bool[] freeMins = Array.Empty<bool>();
  private int[] queenRows = Array.Empty<int>();

  public override object Execute()
  {
    bool result = true;
    for (int i = 0; i < 10; i++)
    {
      result = result && _queens();
    }

    return result;
  }

  private bool _queens()
  {
    freeRows = new bool[8];
    Array.Fill(freeRows, true);
    freeMaxs = new bool[16];
    Array.Fill(freeMaxs, true);
    freeMins = new bool[16];
    Array.Fill(freeMins, true);
    queenRows = new int[8];
    Array.Fill(queenRows, -1);

    return PlaceQueen(0);
  }

  private bool PlaceQueen(int c)
  {
    for (int r = 0; r < 8; r++)
    {
      if (GetRowColumn(r, c))
      {
        queenRows[r] = c;
        SetRowColumn(r, c, false);

        if (c == 7)
        {
          return true;
        }

        if (PlaceQueen(c + 1))
        {
          return true;
        }

        SetRowColumn(r, c, true);
      }
    }

    return false;
  }

  private bool GetRowColumn(int r, int c)
  {
    return freeRows[r] && freeMaxs[c + r] && freeMins[c - r + 7];
  }

  private void SetRowColumn(int r, int c, bool v)
  {
    freeRows[r] = v;
    freeMaxs[c + r] = v;
    freeMins[c - r + 7] = v;
  }

  public override bool VerifyResult(object result)
  {
    return (bool) result;
  }
}